rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,258 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 126 - Backend Migrations
+
+## Required Approvers
+
+- Engineering @zmb3 && (@fspmarshall || @espadollini )
+- Product: ( @klizhentas || @russjones )
+
+## What
+
+A more reliable and robust mechanism to perform backend migrations.
+
+## Why
+
+The upgrade process requires scaling Auth down to a single instance to ensure that migrations are only performed once as well as preventing a new version and an old version from operating on the same keys with different schema. This is cumbersome and makes the upgrade process a pain point for cluster admins. Scaling Auth down and back up also results in an [uneven load](https://github.com/gravitational/teleport/issues/7029) which can cause connectivity issues and backend latency that can result in a thundering herd.
+
+Migrations to date have existed as a function that is called during Auth initialization. The migration function is left as is for a release or two and is then deleted. This makes tracking when and which migrations have been applied difficult and prevents skipping any major versions when upgrading since there are no guarantees that the latest version will include all the migrations between versions.
+
+## Details
+
+### When is a Migration required?
+
+Migrations are not needed for every change to a backend resource. Adding a new field to the resource should be a backward compatible operation by nature of Protocol Buffers and does not require an actual migration. Any major changes to a resource that alter its shape, change its encoding, or moving a single resource into several other resources are candidates for a migration.
```

Comment:
How to deal with this situation then?

Auth server A is on old version
Auth server B is on new version (adds a field).

Auth server B writes a resource
Auth server A reads a resource, skips the field, returns to client
Client updates the resource, overwrites the field

Reply:
@rosstimothy In practice, we would have to add new fields all the time. So instead, we can put restrictions on upgrade procedure instead of limiting ourselves to only major versions.